Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift

```diff
@@ -91,24 +92,76 @@ public protocol SyntaxExpressibleByStringInterpolation:
     where Self.StringInterpolation == SyntaxStringInterpolation {
   /// Create an instance of this syntax node by parsing it from the given
   /// parser.
-  static func parse(from parser: inout Parser) -> Self
+  static func parse(from parser: inout Parser) throws -> Self
+}
+
+enum SyntaxStringInterpolationError: Error, CustomStringConvertible {
+  case didNotConsumeAllTokens(remainingTokens: [TokenSyntax])
+  case producedInvalidNodeType(expectedType: SyntaxProtocol.Type, actualType: SyntaxProtocol.Type)
+  case diagnostics([Diagnostic])
+
+  var description: String {
+    switch self {
+    case .didNotConsumeAllTokens(remainingTokens: let tokens):
+      return "Extraneous text in snippet: '\(tokens.map(\.description).joined())'"
+    case .producedInvalidNodeType(expectedType: let expectedType, actualType: let actualType):
+      return "Parsing the code snippet was expected to produce a \(expectedType) but produced a \(actualType)"
+    case .diagnostics(let diagnostics):
+      return diagnostics.map(\.debugDescription).joined(separator: "\n")
+    }
+  }
 }
 
 extension SyntaxExpressibleByStringInterpolation {
   /// Initialize a syntax node by parsing the contents of the interpolation.
+  /// This function is marked `@_transparent` so that fatalErrors raised here
+  /// are reported at the string literal itself.
+  /// This makes debugging easier because Xcode will jump to the string literal
+  /// that had a parsing error instead of the initializer that raised the `fatalError`
+  @_transparent
   public init(stringInterpolation: SyntaxStringInterpolation) {
```

Comment:
Does anyone have any concerns about this `@_transparent` hack? @DougGregor

Reply:
Can we do this using default arguments with `#file` and `#line` instead? `@_transparent` is a messy feature that we'd prefer not spread much farther.

Reply:
No, because Xcode’s debugger will still jump to the position where the `fatalError` was called without `@_transparent`. 😬